src/rgw/rgw_bucket.cc

```diff
@@ -2130,34 +2133,462 @@ void rgw_data_change_log_entry::decode_json(JSONObj *obj) {
+      entries.push_back(std::move(log_entry));
+    }
+    if (truncated)
+      *truncated = !more;
+    if (out_marker && !log_entries.empty()) {
+      *out_marker = log_entries.back().marker;
+    }
+    return 0;
+  }
+  int get_info(int index, RGWDataChangesLogInfo *info) override {
+    try {
+      auto& fifo = fifos[index];
+      fifo->read_meta(ca::use_blocked);
+      auto m = fifo->meta();
+      auto p = m.head_part_num;
+      if (p < 0) {
+	info->marker = neorados::cls::fifo::marker{}.to_string();
+      info->last_update = ceph::real_clock::zero();
+      return 0;
+      }
+      auto h = fifo->get_part_info(p, ca::use_blocked);
+      info->marker = neorados::cls::fifo::marker{p, h.max_ofs}.to_string();
+      info->last_update = h.max_time;
+    } catch (const bs::system_error& e) {
+      lderr(cct) << "RGWDataChangesFIFO::get_info(): " << e.what() << dendl;
+      return ceph::from_error_code(e.code());
+    }
+    return 0;
+  }
+  int trim(int index, std::string_view marker) override {
+    bs::error_code ec;
+    fifos[index]->trim(marker, ca::use_blocked[ec]);
+    return ceph::from_error_code(ec);
+  }
+  int trim(int index, std::string_view marker,
+	   librados::AioCompletion* c) override {
+    fifos[index]->trim(marker, c);
+    return 0;
+  }
+  std::string_view max_marker() const override {
+    static const std::string mm =
+      neorados::cls::fifo::marker::max().to_string();
+    return std::string_view(mm);
+  }
+};
+
+class RGWDataChangesLegacyFIFO final : public RGWDataChangesBE {
+  using centries = std::vector<ceph::buffer::list>;
+  std::vector<std::unique_ptr<rgw::cls::fifo::FIFO>> fifos;
+public:
+  RGWDataChangesLegacyFIFO(CephContext* cct, librados::Rados* rados,
+			   const rgw_pool& log_pool)
+    : RGWDataChangesBE(cct) {
+    librados::IoCtx ioctx;
+    auto num_shards = cct->_conf->rgw_data_log_num_shards;
+    auto r = rgw_init_ioctx(rados, log_pool.name, ioctx,
+			    true, false);
+    if (r < 0) {
+      throw bs::system_error(ceph::to_error_code(r));
+    }
+    fifos.reserve(num_shards);
+    for (auto i = 0; i < num_shards; ++i) {
+      r = rgw::cls::fifo::FIFO::create(ioctx, get_oid(i),
+				       &fifos[i], null_yield);
+      if (r < 0) {
+	throw bs::system_error(ceph::to_error_code(r));
+      }
+    }
+  }
+  ~RGWDataChangesLegacyFIFO() override = default;
+  static bool exists(CephContext* cct, librados::Rados* rados,
+		     const rgw_pool& log_pool) {
+    auto num_shards = cct->_conf->rgw_data_log_num_shards;
+    librados::IoCtx ioctx;
+    auto r = rgw_init_ioctx(rados, log_pool.name, ioctx,
+			   false, false);
+    if (r < 0) {
+      if (r == bs::errc::no_such_file_or_directory) {
+	return false;
+      } else {
+	throw std::system_error(ceph::to_error_code(r));
+      }
+    }
+    for (auto i = 0; i < num_shards; ++i) {
+      std::unique_ptr<rgw::cls::fifo::FIFO> fifo;
+      auto r = rgw::cls::fifo::FIFO::open(ioctx, get_oid(cct, i),
+					  &fifo, null_yield);
```

Comment:
nautilus doesn't have yield context--does that matter?

Reply:
No, we've got a nautilus compatible version in legacy.